Remotion/Data/DomainObjects/Mapping/TypeDefinitionCollectionFactory.cs

```diff
@@ -0,0 +1,111 @@
+// This file is part of the re-motion Core Framework (www.re-motion.org)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+// 
+// The re-motion Core Framework is free software; you can redistribute it 
+// and/or modify it under the terms of the GNU Lesser General Public License 
+// as published by the Free Software Foundation; either version 2.1 of the 
+// License, or (at your option) any later version.
+// 
+// re-motion is distributed in the hope that it will be useful, 
+// but WITHOUT ANY WARRANTY; without even the implied warranty of 
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Remotion.Data.DomainObjects.Mapping.Builder;
+using Remotion.Utilities;
+
+namespace Remotion.Data.DomainObjects.Mapping
+{
+  /// <summary>
+  /// The <see cref="TypeDefinitionCollectionFactory"/> is used to get the list of <see cref="TypeDefinition"/>s for a set of types.
+  /// It automatically sets the relationships between the different <see cref="TypeDefinition"/>s.
+  /// </summary>
+  public class TypeDefinitionCollectionFactory
+  {
+    private readonly IMappingObjectFactory _mappingObjectFactory;
+
+    public TypeDefinitionCollectionFactory (IMappingObjectFactory mappingObjectFactory)
+    {
+      ArgumentUtility.CheckNotNull("mappingObjectFactory", mappingObjectFactory);
+
+      _mappingObjectFactory = mappingObjectFactory;
+    }
+
+    public TypeDefinition[] CreateTypeDefinitionCollection (IEnumerable<Type> types)
+    {
+      ArgumentUtility.CheckNotNull("types", types);
+
+      var builderNodesLookup = new Dictionary<Type, TypeDefinitionBuilderNode>();
+
+      foreach (var type in types)
+        AddOrCreateBuilderNode(builderNodesLookup, type);
+
+      var leafNodes = builderNodesLookup.Values
+          .Where(e => e.IsLeafNode)
+          .OfType<ClassDefinitionBuilderNode>()
+          .ToArray();
+
+      foreach (var leafNode in leafNodes)
+        leafNode.BuildTypeDefinition(_mappingObjectFactory);
+
+      var results = new List<TypeDefinition>();
+      foreach (var builderNode in builderNodesLookup.Values)
+      {
+        builderNode.SetReadOnly();
```

Comment:
Because the building logic needs to go from leaf nodes upwards. That simplifies the construction while also filtering out nodes that should not get a class definition. For example, the inheritance root only makes sense going from bottom up. If we were to build every node the logic would also need to check if the root is below. This is just more complicated that having the bit of extra logic here.

Reply:
I see. should we add a comment regarding this particular detail? First, we should resolve the isbuilt/isreadonly duality and go from there?